core/utils/historical-data/src/linear.rs

```diff
@@ -0,0 +1,428 @@
+//! could be use).
+//! Access to latest value in history should be the less costy access.
+
+#[cfg(not(feature = "std"))]
+use rstd::{vec::Vec, vec};
+use rstd::marker::PhantomData;
+use rstd::borrow::Cow;
+use crate::HistoricalValue;
+
+
+/// Array like buffer for in memory storage.
+/// By in memory we expect that this will
+/// not required persistence and is not serialized.
+#[cfg(not(feature = "std"))]
+pub(crate) type InMemory<V, I> = Vec<HistoricalValue<V, I>>;
+
+/// Array like buffer for in memory storage.
+/// By in memory we expect that this will
+/// not required persistence and is not serialized.
+#[cfg(feature = "std")]
+pub(crate) type InMemory<V, I> = smallvec::SmallVec<[HistoricalValue<V, I>; ALLOCATED_HISTORY]>;
+
+/// Size of preallocated history per element.
+/// Currently at two for committed and prospective only.
+/// It means that using transaction in a module got a direct allocation cost.
+#[cfg(feature = "std")]
+const ALLOCATED_HISTORY: usize = 2;
+
+/// Arraylike buffer with in place byte data.
+/// Can be written as is in underlying storage.
+/// Could be extended to direct access memory too.
+#[derive(Debug, Clone)]
+pub struct Serialized<'a, F>(SerializedBuff<'a>, PhantomData<F>);
+
+impl<'a, 'b, F> PartialEq<Serialized<'b, F>> for Serialized<'a, F> {
+  fn eq(&self, other: &Serialized<'b, F>) -> bool {
+		self.0.eq(&other.0)
+	}
+}
+
+impl<'a, F> Eq for Serialized<'a, F> { }
+
+/// Internal buffer, it is either a readonly view other the
+/// serialized data or the pending changes.
+type SerializedBuff<'a> = Cow<'a, [u8]>;
+
+/// Serialized specific behavior.
+pub trait SerializedConfig {
+	/// Encoded empty slice.
+	fn empty() -> &'static [u8];
+	/// Size needed to encode version.
+	/// Should be a static value.
+	fn version_len() -> usize;
+}
+
+#[derive(Debug, Clone)]
+#[cfg_attr(any(test, feature = "test"), derive(PartialEq))]
+/// Serialize without versioning.
+pub struct NoVersion;
+
+#[derive(Debug, Clone)]
+#[cfg_attr(any(test, feature = "test"), derive(PartialEq))]
+/// Serialize with default version.
+pub struct DefaultVersion;
+
+impl SerializedConfig for NoVersion {
+	fn empty() -> &'static [u8] {
+		&EMPTY_SERIALIZED
+	}
+	fn version_len() -> usize {
+		0
+	}
+}
+
+impl SerializedConfig for DefaultVersion {
+	fn empty() -> &'static [u8] {
+		&DEFAULT_VERSION_EMPTY_SERIALIZED
+	}
+	fn version_len() -> usize {
+		1
+	}
+}
+
+// Length in number of bytes for an encoded size.
+// Current value is one of a u64.
+// Used both for number of element in history and
+// length of elements.
+const SIZE_BYTE_LEN: usize = 8;
+
+// Basis implementation must be on par with InMemory.
+// Those method could be move to a 'VecLike' trait.
+//
+// Those function requires prior index checking.
+impl<'a, F: SerializedConfig> Serialized<'a, F> {
+
+	pub fn into_owned(self) -> Serialized<'static, F> {
+		Serialized(Cow::from(self.0.into_owned()), PhantomData)
+	}
```

Comment:
also this function seem never used, nor it is documented, also Serialized implements Clone and the only difference between this and clone is that internally the value is owned instead of borrowed but usre of linear don't access this so from his point of view there is no difference